### ISSUE TITLE
Race condition in in Deno.serve() and fetch()

### ISSUE THREAD

**Issue by bartlomieju** [denoland/deno#20443]
Extracted from `cli/tests/unit/serve_test.ts`:

```js
import { assertEquals, fail } from "./test_util/std/testing/asserts.ts";
import { deferred } from "./test_util/std/async/mod.ts";
const servePort = 4502;
const testFn = [];

function onListen<T>(
  p: Deferred<T>,
): ({ hostname, port }: { hostname: string; port: number }) => void {
  return () => {
    p.resolve();
  };
}
function createOnErrorCb(ac: AbortController): (err: unknown) => Response {
  return (err) => {
    console.error(err);
    ac.abort();
    return new Response("Internal server error", { status: 500 });
  };
}

function createStreamTest(count: number, delay: number, action: string) {
  function doAction(controller: ReadableStreamDefaultController, i: number) {
    if (i == count) {
      if (action == "Throw") {
        controller.error(new Error("Expected error!"));
      } else {
        controller.close();
      }
    } else {
      controller.enqueue(`a${i}`);

      if (delay == 0) {
        doAction(controller, i + 1);
      } else {
        setTimeout(() => doAction(controller, i + 1), delay);
      }
    }
  }

  function makeStream(_count: number, delay: number): ReadableStream {
    return new ReadableStream({
      start(controller) {
        if (delay == 0) {
          doAction(controller, 0);
        } else {
          setTimeout(() => doAction(controller, 0), delay);
        }
      },
    }).pipeThrough(new TextEncoderStream());
  }

  const fn = async () => {
    const ac = new AbortController();
    const listeningPromise = deferred();
    const server = Deno.serve({
      handler: (_request) => {
        return new Response(makeStream(count, delay));
      },
      port: servePort,
      signal: ac.signal,
      onListen: onListen(listeningPromise),
      onError: createOnErrorCb(ac),
    });
    console.log("started server", count, delay, action);

    try {
      await listeningPromise;
      const resp = await fetch(`http://127.0.0.1:${servePort}/`);
      if (action == "Throw") {
        try {
          await resp.text();
          fail();
        } catch (_) {
          // expected
        }
      } else {
        const text = await resp.text();

        let expected = "";
        for (let i = 0; i < count; i++) {
          expected += `a${i}`;
        }

        assertEquals(text, expected);
      }
    } finally {
      ac.abort();
      await server.finished;
      console.log("waited server", count, delay, action);
    }
  };
  Object.defineProperty(fn, "name", {
    value: `httpServerStreamCount${count}Delay${delay}${action}`,
  });
  testFn.push(fn);
}

for (const count of [0, 1, 2, 3]) {
  for (const delay of [0, 1, 25]) {
    // Creating a stream that errors in start will throw
    if (delay > 0) {
      createStreamTest(count, delay, "Throw");
    }
    createStreamTest(count, delay, "Close");
  }
}

for (const fn of testFn) {
  await fn();
}

```

```
deno run -A test.ts
error: Uncaught (in promise) TypeError: error sending request for url (http://127.0.0.1:4502/): connection closed before message completed
      const resp = await fetch(`http://127.0.0.1:${servePort}/`);
                   ^
    at async mainFetch (ext:deno_fetch/26_fetch.js:277:12)
    at async fetch (ext:deno_fetch/26_fetch.js:501:7)
    at async httpServerStreamCount0Delay1Throw (file:///Users/ib/dev/deno/test.ts:68:20)
    at async file:///Users/ib/dev/deno/test.ts:109:3
```

I found this when working on https://github.com/denoland/deno/issues/20438; it appears it currently works fine with `deno test` due to interaction with op sanitizer that adds a 1ms delay before running subsequent test.

**Comment by bartlomieju** [comment]
Adding `new Promise(resolve => setTimeout(resolve, 1))` after `await fn()` makes all functions pass properly.

**Comment by lucacasonato** [comment]
Fixed by #20680
